### PR TITLE
Users and groups management APIs for OpenLDAP

### PR DESCRIPTION
**Pull request by DavidePrincipi**
Implement users and group management APIs for a LDAP RFC2307 DB schema.

## Adding

- add-user, alter-user, remove-user 
- add-group, alter-group, remove-group

User attributes:
- password, 
- full name (displayName),
- ~~locked state~~,
- group membership,
- uidNumber, gidNumber (randomly assigned)

Group attributes:
- description,
- members,
- gidNumber

## Testing

Install the module from `users-api` branch

    api-cli run add-internal-provider --data '{"image":"ghcr.io/nethserver/openldap:users-api","node":1}'
    api-cli run module/openldap1/configure-module --data '{"provision":"new-domain","admuser":"admin","admpass":"secret","domain":"dom.test"}'

## Postponed
* uidNumber and gidNumber are assigned randomly: the allocation procedure must be improved
* Domain password policy management
* Add "password does not expire" user flag
* Account locked state

## See also

https://github.com/NethServer/ns8-core/pull/241 for the read side.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -55,4 +55,22 @@ are sent to stderr, which is forwarded to Systemd journal. Set
...
+
+    api-cli run module/openldap1/alter-group --data '{"group":"mygroup1","description":"My Group 1"}'
+
+Create user `first.user` as member of `mygroup1`
+
+    api-cli run module/openldap1/add-user --data '{"user":"first.user","display_name":"First User","password":"Nethesis,1234","groups":["mygroup1"]}'
+
+Change First User's password
+
+    api-cli run module/openldap1/alter-user --data '{"user":"usr1","password":"Neth,123"}'
```
Comment (gsanchietti) [inline README.md:76]:
```suggestion
    api-cli run module/openldap1/alter-user --data '{"user":"first.user","password":"Neth,123"}'
```

File: server/usr/local/bin/add-group [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    printf "Usage: %s [OPTIONS] GROUP\n" "$0"
+    printf "    GROUP                    Add/Create a new GROUP of users\n"
+    printf "    OPTIONS\n"
+    printf "      -u user1,user2,...    Comma-separated list of initial group members\n"
+    printf "      -d description        The group description\n"
+    printf "      -g numgid             Set the numeric group identifier\n"
+    exit 2
+}
+
+# XXX default ngid? e.g. max uidNumber + 1
```
Comment (gsanchietti) [inline server/usr/local/bin/add-group:38]: Why use a random GID? Maybe it's better to get current maximum GID and add 1.

File: server/usr/local/bin/add-user [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    printf "      -d display_name       The user display name\n"
+    printf "      -p password           Initial password; \"-\" reads from standard input\n"
+    printf "      -G numgid             Set the primary group number\n"
+    printf "      -U numuid             Set the numeric user identifier\n"
+    exit 2
+}
+
+ngid=1001
+# XXX default nuid? e.g. max uidNumber + 1
+nuid=$(( 1003 + RANDOM % 8996 ))
```
Comment (gsanchietti) [inline server/usr/local/bin/add-user:42]: Why use a random UID? Maybe it's better to get current maximum UID and add 1.

File: server/usr/local/bin/add-user [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+done
+shift $((OPTIND - 1))
+user=${1}
+if [ -z "${user}" ]; then
+    usage_error "Missing USER argument"
+fi
+
+if [ "${pflag}" = 1 ]; then
+    if [ "${password}" = '-' ]; then
+        echo -n "Enter password: "
```
Comment (gsanchietti) [inline server/usr/local/bin/add-user:63]: Not sure this is needed: the script should not be called from CLI. Am I wrong?

File: server/usr/local/bin/alter-user [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+errors=0
+
+if [ "${pflag}" = 1 ]; then
+    if [ "${password}" = '-' ]; then
+        echo -n "Enter password: "
+        read -r password # Read password from stdin
+        echo
+    elif [ "${password}" = '' ]; then
+        password=$(dd if=/dev/urandom of=/dev/stdout bs=32 count=1 | base64 -w 0)
```
Comment (gsanchietti) [inline server/usr/local/bin/alter-user:61]:
```suggestion
        password=$(echo $RANDOM | base64 -w 0)
```

File: server/usr/local/bin/add-group (hunk 2 above)
Comment (DavidePrincipi) [reply server/usr/local/bin/add-group:38]: uidNumber and gidNumber should not be important for ns8 apps. I decided to postpone this issue after considering the users import script implementation and measuring performances with 10k users.

File: server/usr/local/bin/add-user (hunk 3 above)
Comment (DavidePrincipi) [reply server/usr/local/bin/add-user:42]: See https://github.com/NethServer/ns8-openldap/pull/3#discussion_r866559687

File: server/usr/local/bin/add-user (hunk 4 above)
Comment (DavidePrincipi) [reply server/usr/local/bin/add-user:63]: Reading the secret from a file descriptor is 
This is similar to Samba change password script too.

File: server/usr/local/bin/alter-user (hunk 5 above)
Comment (DavidePrincipi) [reply server/usr/local/bin/alter-user:61]: I'd prefer a bit more (pseudo) entropy :man_shrugging:

File: server/usr/local/bin/add-user (hunk 4 above)
Comment (gsanchietti) [reply server/usr/local/bin/add-user:63]: I'm not arguing about reading the pass from stdin.
I just noted that the user will never see "Enter password: " prompt.

File: server/usr/local/bin/alter-user (hunk 5 above)
Comment (gsanchietti) [reply server/usr/local/bin/alter-user:61]: I'd prefer simpler code :)

File: server/usr/local/bin/add-group (hunk 2 above)
Comment (gsanchietti) [reply server/usr/local/bin/add-group:38]: I still do not like a random UID/GID

**Pull request merged**